### PR TITLE
Add WatchForObject to e2e test framework

### PR DESCRIPTION
**Pull request by karlkfi**

### CODE REVIEW

File: e2e/nomostest/wait_for_sync.go [hunk 1]
```diff
@@ -476,7 +476,9 @@ func (nt *NT) WaitForRepoSyncStalledError(rsNamespace, rsName, reason, message s
 // 	return nil
 // }
 
-func validateError(errs []v1beta1.ConfigSyncError, code string, message string) error {
+// ValidateConfigSyncError returns true if the specified errors contain an error
+// with the specified error code and (partial) message.
+func ValidateConfigSyncError(errs []v1beta1.ConfigSyncError, code string, message string) error {
```
Comment (sdowell) [inline e2e/nomostest/wait_for_sync.go:481]: Why the export and name change?

Comment (karlkfi) [reply e2e/nomostest/wait_for_sync.go:481]: Wanted to use it for custom predicates. Tho it turns out i didn't need to after I moved the predicates into the nomostest package. I could change it back...

File: e2e/nomostest/predicates.go [hunk 2]
```diff
@@ -594,3 +595,75 @@ func RoleBindingHasName(expectedName string) Predicate {
...
+		if !ok {
+			return WrongTypeErr(o, &v1beta1.RootSync{})
+		}
+		return ValidateConfigSyncError(rs.Status.Source.Errors, errCode, errMessage)
+	}
+}
+
+// RepoSyncHasSourceError returns an error if the RootSync does not have the
+// specified Source error code and (optional, partial) message.
+func RepoSyncHasSourceError(nt *NT, errCode string, errMessage string) Predicate {
```
[the remark this answers is not in the document]
Comment (karlkfi) [reply e2e/nomostest/predicates.go:613]: I didn't want to refactor everything to use them in the same PR. I just wanted to replace the WaitForObject I added recently. So yes, incremental.

[the remark this answers is not in the document]
Comment (karlkfi) [reply e2e/nomostest/predicates.go:613]: I'll replace all the old WaitFor*Error funcs in a follow up

[the remark this answers is not in the document]
Comment (sdowell) [reply e2e/nomostest/predicates.go:613]: Sounds good, just wanted to clarify

File: e2e/nomostest/wait_for_sync.go (hunk 1 above)
Comment (sdowell) [reply e2e/nomostest/wait_for_sync.go:481]: Might as well not change it if there's no need 🤷

File: e2e/nomostest/predicates.go [hunk 3]
```diff
@@ -594,3 +595,75 @@ func RoleBindingHasName(expectedName string) Predicate {
 		return nil
 	}
 }
+
+// RootSyncHasSourceError returns an error if the RootSync does not have the
+// specified Source error code and (optional, partial) message.
+func RootSyncHasSourceError(nt *NT, errCode string, errMessage string) Predicate {
```
[the remark this answers is not in the document]
Comment (karlkfi) [reply e2e/nomostest/predicates.go]: fixed.

File: e2e/nomostest/wait_for_sync.go (hunk 1 above)
Comment (karlkfi) [reply e2e/nomostest/wait_for_sync.go]: reverted

File: e2e/nomostest/wait_for_sync.go [hunk 4]
```diff
@@ -444,7 +444,7 @@ func (nt *NT) WaitForRepoSyncStalledError(rsNamespace, rsName, reason, message s
 
 // TODO: Uncomment when Syncing condition consistency is fixed
 // func validateRootSyncError(statusErrs []v1beta1.ConfigSyncError, syncingCondition *v1beta1.RootSyncCondition, code string, message string, expectedErrorSourceRefs []v1beta1.ErrorSource) error {
-// 	if err := validateError(statusErrs, code, message); err != nil {
+// 	if err := ValidateConfigSyncError(statusErrs, code, message); err != nil {
```
[the remark this answers is not in the document]
Comment (karlkfi) [reply e2e/nomostest/wait_for_sync.go]: done

**Pull request merged**